### DISCUSSION TITLE
How to set the Field complexity in ComplexityValidationRule with v8?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [graphql-dotnet/graphql-dotnet#4069] [Q&A] [answered]
In v7, this is how we set the `FieldImpact` settings. `FieldImpact` was nullable.

```
.Concat(context.RequestServices.GetServices<IValidationRule>())
.Append(new ComplexityValidationRule(new ComplexityConfiguration
{
    MaxDepth = _settings.MaxDepth,
    MaxComplexity = _settings.MaxComplexity,
    FieldImpact = _settings.FieldImpact
}));
```

Will the correct migration path be this?

```
.Concat(context.RequestServices.GetServices<IValidationRule>())
.Append(new ComplexityValidationRule(new ComplexityOptions
{
    MaxDepth = _settings.MaxDepth,
    MaxComplexity = _settings.MaxComplexity,
    DefaultComplexityImpactDelegate = (ctx) =>
    {
        return new FieldComplexityResult()
        {
            FieldImpact = _settings.FieldImpact ?? 0,
        };
    }
}));
```

Also, when migration the `IValidationRule` implementation, would the old `ValidateAsync` method become  `GetPreNodeVisitorAsync` `GetVariableVisitorAsync`,  or `GetPostNodeVisitorAsync`. I am not familar with the library enough to determine how to correctly migrate this.

**Comment by Shane32** [comment]
To use the old validation rule (to be removed in v9):
```cs
.Concat(context.RequestServices.GetServices<IValidationRule>())
.Append(new LegacyComplexityValidationRule(new LegacyComplexityConfiguration
{
    MaxDepth = _settings.MaxDepth,
    MaxComplexity = _settings.MaxComplexity,
    FieldImpact = _settings.FieldImpact
}));
```

The new validation rule computes complexity and depth quite differently and has three different default values:
- `DefaultScalarImpact`
- `DefaultObjectImpact`
- `DefaultListImpactModifier`

Please read https://graphql-dotnet.github.io/docs/guides/complexity-analyzer/ for details on how the complexity is computed, explaining how you may wish to set these defaults.  For instance, perhaps this is most useful:

```cs
.Concat(context.RequestServices.GetServices<IValidationRule>())
.Append(new ComplexityValidationRule(new ComplexityOptions
{
    MaxDepth = _settings.MaxDepth,
    MaxComplexity = _settings.MaxComplexity,
    DefaultObjectImpact = _settings.FieldImpact ?? 1,
}));
```

I do not have an equivalent `DefaultComplexityImpactDelegate` implementation to simulate the old behavior.

**Comment by Shane32** [comment]
> would the old `ValidateAsync` method become `GetPreNodeVisitorAsync`...

yes, it would become `GetPreNodeVisitorAsync` for equivalent behavior

**Comment by MikeAlhayek** [comment]
@Shane32 thank you for your feedback! I am trying to upgrade OrchardCore to use version 8.0.2 from 7.8.0. Running into an issue and not sure if you have any insight. Couple of my unit tests are failing with the following exception:

```
System.AggregateException: 'The schema is invalid. See inner exceptions for details. (The field 'Amount' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.) (The input field 'Amount' of an Input Object 'PricePartWhereInput' must be an input type.) (The field 'Amount_not' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.) (The input field 'Amount_not' of an Input Object 'PricePartWhereInput' must be an input type.) (The field 'Amount_in' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.) (The input field 'Amount_in' of an Input Object 'PricePartWhereInput' must be an input type.) (The field 'Amount_not_in' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.) (The input field 'Amount_not_in' of an Input Object 'PricePartWhereInput' must be an input type.) (The field 'Amount_gt' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.) (The input field 'Amount_gt' of an Input Object 'PricePartWhereInput' must be an input type.) (The field 'Amount_gte' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.) (The input field 'Amount_gte' of an Input Object 'PricePartWhereInput' must be an input type.) (The field 'Amount_lt' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.) (The input field 'Amount_lt' of an Input Object 'PricePartWhereInput' must be an input type.) (The field 'Amount_lte' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.) (The input field 'Amount_lte' of an Input Object 'PricePartWhereInput' must be an input type.) (The field 'Amount' of an Object type 'PricePart' must have non-null 'ResolvedType' property for all types in the chain.) (The field 'Amount' of an Object type 'PricePart' must be an output type.)'
```

As far as I can tell, this is the method that provides the schema 

https://github.com/OrchardCMS/OrchardCore/blob/e9c183ffbf8aa0f3ddb2e13e8b31f639150fccc4/src/OrchardCore.Modules/OrchardCore.ContentFields/GraphQL/Fields/ContentFieldsProvider.cs#L101-L128


Here is where the filter fields
https://github.com/OrchardCMS/OrchardCore/blob/e9c183ffbf8aa0f3ddb2e13e8b31f639150fccc4/src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/Queries/WhereInputObjectGraphType.cs#L90C13-L91

https://github.com/OrchardCMS/OrchardCore/blob/e9c183ffbf8aa0f3ddb2e13e8b31f639150fccc4/src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/Queries/WhereInputObjectGraphType.cs#L120-L125

Perhaps something else changes iv v8 that is causing this issue. Any clues you may be able to share?

The `Amount` field which is in the exception is `DecimalGraphType`

**Reply by Shane32** [reply]
First, let's write out all the schema validation errors encountered:

- The field 'Amount' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.
- The input field 'Amount' of an Input Object 'PricePartWhereInput' must be an input type.
- The field 'Amount_not' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.
- The input field 'Amount_not' of an Input Object 'PricePartWhereInput' must be an input type.
- The field 'Amount_in' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.
- The input field 'Amount_in' of an Input Object 'PricePartWhereInput' must be an input type.
- The field 'Amount_not_in' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.
- The input field 'Amount_not_in' of an Input Object 'PricePartWhereInput' must be an input type.
- The field 'Amount_gt' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.
- The input field 'Amount_gt' of an Input Object 'PricePartWhereInput' must be an input type.
- The field 'Amount_gte' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.
- The input field 'Amount_gte' of an Input Object 'PricePartWhereInput' must be an input type.
- The field 'Amount_lt' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.
- The input field 'Amount_lt' of an Input Object 'PricePartWhereInput' must be an input type.
- The field 'Amount_lte' of an Input Object type 'PricePartWhereInput' must have non-null 'ResolvedType' property for all types in the chain.
- The input field 'Amount_lte' of an Input Object 'PricePartWhereInput' must be an input type.
- The field 'Amount' of an Object type 'PricePart' must have non-null 'ResolvedType' property for all types in the chain.
- The field 'Amount' of an Object type 'PricePart' must be an output type.

Now, GraphQL in `SchemaValidationVisitor.cs:326` is going to ensure that `field.ResolvedType != null` after unwrapping any list/non-null wrappers.  So it must be seeing that `field.ResolvedType` is `null` somehow.  So then it would make sense that it does not recognize the field as an input type.  So it is apparent that all your errors relate to a null `ResolvedType` property on the following fields:

- `PricePartWhereInput.Amount` and `Amount...` fields
- `PricePart.Amount`

Interestingly one is an input type and one is an output type.  Also, `SchemaTypes` should initialize all the `ResolvedType` fields during schema initialization prior to schema validation.  I'm not sure how they could be `null`.

Do you have your progress in a PR?  Perhaps I could simply download the repo and run the unit tests, and see if I can reproduce the error on my end.

**Reply by Shane32** [reply]
I also doubt we changed those schema validation rules for v8, so it must be some other change that is causing the schema to leave those specific `ResolvedType` properties as `null`.

**Reply by MikeAlhayek** [reply]
@Shane32 thank you very much for this. Yes the PR is here https://github.com/OrchardCMS/OrchardCore/pull/16736

**Reply by Shane32** [reply]
Well, the reason that `ResolvedType` is `null` is partially because `DynamicPartGraphType` has its fields added during the `Initialize` method rather than within the constructor, which occurs after `SchemaTypes` initializes the schema.

**Reply by Shane32** [reply]
Here is where the functionality was changed: https://github.com/graphql-dotnet/graphql-dotnet/pull/3761/files#diff-a2e950d57fa47ff02d0d34385271cecef99cd9074cf102f47fba117dc69e605fR334-R337

**Reply by MikeAlhayek** [reply]
@Shane32 thank you again! Is this an issue that was introduced in v8 that will be fixed, or is it something we need to handle differently on our end? I would expect the initialize method to be called before the validation.

**Reply by Shane32** [reply]
I don't think that changing it back is an option.  The ContentFieldsProvider.cs link you posted indeed is where the field type is generated without the `ResolvedType` property being set.  Normally schema initialization would take care of this, but not when it's added within the `Initialize` method rather than the constructor.  A different approach will have to be considered.

1. Just move the code from within `DynamicPartGraphType.Initialize` into the constructor.  Need to consider why it was in the `Initialize` method in the first place.  Right now it's using a self-activated service provider, and `ISchema` is implementing `IServiceProvider` which is returning the same instance.  So, "DI" might happily inject the service provider with no effective change.
2. Use an `AsyncLocal` to store the `schema` instance and move the initialization code into the constructor.  This should perform exactly as before.

**Reply by Shane32** [reply]
Option 1 is problematic because the code uses `ISchema` and not just `IServiceProvider`.

Option 3: change GraphQL or write a custom schema provider implementation so that `ISchema` can be pulled from the service provider when using a self-activated schema.

**Reply by Shane32** [reply]
> @Shane32 thank you again! Is this an issue that was introduced in v8 that will be fixed, or is it something we need to handle differently on our end? I would expect the initialize method to be called before the validation.

Initialize is currently called before validation, but after initialization.  Previously it was called prior to initialization and validation.

**Reply by Shane32** [reply]
With `Initialize` as it was in v7, `schema` does not have a references to all types that will be initialized by GraphQL.NET.  However, it is free to add fields and etc.

With `Initialize` in its new location, now `Initialize` has a fully initialized `schema` instance to access, but it must leave the schema in an initialized state when finished.

Interestingly, your code actually uses a bit of both -- both reading the schema and adding new fields.  Hmmm.....

**Reply by MikeAlhayek** [reply]
@Shane32 I update the `DynamicPartGraphType` as you can see here https://github.com/OrchardCMS/OrchardCore/pull/16736/files#diff-ba39df7e67225d59fdbd60053c3c801984667e83050304762c496804672146ab

But I am still getting the same exception is the unit tests

**Reply by Shane32** [reply]
It's not too hard to get past this problem.  I made it work but the unit test still fails because the schema name converter is not applied after the schema is initialized.  This also can be fixed.

**Reply by Shane32** [reply]
I can't work on it any more today.  Here's what I've done so far:

![image](https://github.com/user-attachments/assets/af41c1d0-cb15-4ef0-801e-4c95039c8cb6)

![image](https://github.com/user-attachments/assets/c3ad3150-ea45-46ed-b1fb-2397f6cb0f70)

![image](https://github.com/user-attachments/assets/eb79b1d7-651a-4acf-9aa8-777dec306278)

![image](https://github.com/user-attachments/assets/7cd68f91-f30a-4937-ad2b-fd63c9283807)

**Reply by Shane32** [reply]
This is compared to your previous commit, not the latest with the code moved to the ctor.  Maybe that would work but I did not check it out yet.

**Reply by Shane32** [reply]
If I was to work on it further, I'd probably see if the `IGraphType` instance could be passed directly in for better optimization.  And the field name needs to use the name converter which is why the test still fails.

**Reply by Shane32** [reply]
https://github.com/graphql-dotnet/graphql-dotnet/pull/4071

**Reply by MikeAlhayek** [reply]
@Shane32 I am back at this PR hoping to finish it soon and complete migrating to v8. When I try to initalize the schema, I am getting the following error:

```
A different instance of the GraphType 'MenuItemsListQueryObjectType' with the name 'MenuItemsListPart' has already been registered within the schema. Please use the same instance for all references within the schema, or use GraphQLTypeReference to reference a type instantiated elsewhere.\r\nTo view additional trace enable GlobalSwitches.TrackGraphTypeInitialization switch.
```

[Here is https://github.com/OrchardCMS/OrchardCore/tree/ma/upgeade-graphql/src/OrchardCore.Modules/OrchardCore.Menu/GraphQL/MenuItemsListQueryObjectType.cs#L6](https://github.com/OrchardCMS/OrchardCore/tree/ma/upgeade-graphql/src/OrchardCore.Modules/OrchardCore.Menu/GraphQL/MenuItemsListQueryObjectType.cs#L6)

It is register here:

https://github.com/OrchardCMS/OrchardCore/tree/ma/upgeade-graphql/src/OrchardCore.Modules/OrchardCore.Menu/GraphQL/Startup.cs#L14

Here is my registration extension https://github.com/OrchardCMS/OrchardCore/tree/ma/upgeade-graphql/src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/ServiceCollectionExtensions.cs#L29

Any idea where I could be registering the same type more than once?

**Reply by Shane32** [reply, marked as answer]
In `MenuItemContentTypeBuilder` you are creating the class directly:

```cs
ResolvedType = new MenuItemsListQueryObjectType(),
```

whereas in `MenuItemInterface` you are not:

```cs
Field<MenuItemsListQueryObjectType>("menuItemsList");
```

So the schema builder is asked to initialize an instance itself + use an already initialized instance, which is not allowed.

The easiest solution is likely to comment out the first reference, as you already have the type specified, letting the schema builder create a single instance itself:

```cs
Type = typeof(MenuItemsListQueryObjectType),
//ResolvedType = new MenuItemsListQueryObjectType(),
```

**Reply by MikeAlhayek** [reply]
@Shane32 Thanks for your help! Was able to complete the migration to 8.2.

**Comment by Shane32** [comment]
Glad it's working!  Open another issue if you have further questions.
